### PR TITLE
Provide a way to register custom DocNode subclasses

### PR DESCRIPTION
**Pull request by pgonzal**
- (API change) Rename `TSDocParserConfiguration` to `TSDocConfiguration`
- (API change) Require an associated `TSDocConfiguration` to be included with the parameters for every `DocNode` constructor
- Introduce `TSDocConfiguration.docNodeManager` for registering custom `DocNode` subclasses and their container relationships

### CODE REVIEW

File: tsdoc/src/nodes/DocNodeContainer.ts [hunk 1]
```diff
@@ -36,24 +36,14 @@ export abstract class DocNodeContainer extends DocNode {
...
-    return false;
-  }
-
   /**
    * Append a node to the container.
    */
   public appendNode(docNode: DocNode): void {
-    if (!this.isAllowedChildNode(docNode)) {
-      throw new Error(`A ${this.kind} cannot contain nodes of type ${docNode.kind}`);
+    if (!this.configuration.docNodeManager.isAllowedChild(this.kind, docNode.kind)) {
```
Comment (iclanton) [inline tsdoc/src/nodes/DocNodeContainer.ts]: >configuration.docNodeManager.isAllowedChild [](start = 14, length = 43)

This is pretty long. Maybe we should alias `isAllowedChild` on the `DocNode` base class.

File: tsdoc/src/configuration/DocNodeManager.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  packageName: string;
+  allowedChildKinds: Set<string>;
+}
+
+/**
+ * Part of the {@link TSDocConfiguration} object.
+ */
+export class DocNodeManager {
+  private static readonly _nodeKindRegExp: RegExp = /^[_a-z][_a-z0-9]*$/i;
+
```
Comment (iclanton) [inline tsdoc/src/configuration/DocNodeManager.ts]: Can you provide a comment describing what this matches?

File: tsdoc/src/configuration/DocNodeManager.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  private readonly _docNodeDefinitionsByKind: Map<string, IRegisteredDocNodeDefinition>
+    = new Map<string, IRegisteredDocNodeDefinition>();
+  private readonly _docNodeDefinitionsByConstructor: Map<DocNodeConstructor, IRegisteredDocNodeDefinition>
+    = new Map<DocNodeConstructor, IRegisteredDocNodeDefinition>();
+
+  public registerDocNodes(packageName: string, definitions: ReadonlyArray<IDocNodeDefinition>): void {
```
Comment (iclanton) [inline tsdoc/src/configuration/DocNodeManager.ts]: >registerDocNodes [](start = 9, length = 16)

Docs?

File: tsdoc/src/configuration/DocNodeManager.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        packageName,
+        allowedChildKinds: new Set<string>()
+      };
+
+      this._docNodeDefinitionsByKind.set(definition.docNodeKind, newDefinition);
+      this._docNodeDefinitionsByConstructor.set(definition.constructor, newDefinition);
+    }
+  }
+
+  public requireRegisteredKind(docNodeKind: string): void {
```
Comment (iclanton) [inline tsdoc/src/configuration/DocNodeManager.ts]: >requireRegisteredKind [](start = 9, length = 21)

Maybe rename this to `throwIfNotRegistered`

File: tsdoc/src/nodes/DocNodeContainer.ts (hunk 1 above)
Comment (pgonzal) [reply tsdoc/src/nodes/DocNodeContainer.ts]: Let's wait and see if any other code actually needs to do this

---
In reply to: [230253352](https://github.com/Microsoft/tsdoc/pull/126#discussion_r230253352) [](ancestors = 230253352)

**Pull request merged**